fatrace.c

```diff
@@ -300,44 +366,83 @@ print_event (const struct fanotify_event_metadata *data,
         event_fd = get_fid_event_fd (data);
 #endif
 
+    bool got_path = false;
+    struct stat st = { .st_uid = -1 };
+    if (option_json && fstat (event_fd, &st) < 0)
+        add_problem("Failed to stat event, cannot determine device or inode.");
     if (event_fd >= 0) {
         /* try to figure out the path name */
         snprintf (printbuf, sizeof (printbuf), "/proc/self/fd/%i", event_fd);
         ssize_t len = readlink (printbuf, pathname, sizeof (pathname));
-        if (len < 0) {
+        if (len >= 0) {
+            pathname[len] = '\0';
+            got_path = true;
+        } else if (option_json) {
+          // no fallback, device/inode will always be printed
+        } else {
             /* fall back to the device/inode */
-            struct stat st;
-            if (fstat(event_fd, &st) < 0) {
+            if (fstat (event_fd, &st) < 0) {
                 warn ("stat");
                 pathname[0] = '\0';
+                add_problem("Failed to stat event, cannot determine device or inode.");
             } else {
                 snprintf (pathname, sizeof (pathname), "device %i:%i inode %ld\n", major (st.st_dev), minor (st.st_dev), st.st_ino);
             }
-        } else {
-            pathname[len] = '\0';
         }
 
         close (event_fd);
     } else {
         snprintf (pathname, sizeof (pathname), "(deleted)");
+        add_problem("Event deleted, cannot determine path.");
     }
 
     /* print event */
     if (option_timestamp == 1) {
         strftime (printbuf, sizeof (printbuf), "%H:%M:%S", localtime (&event_time->tv_sec));
-        printf ("%s.%06li ", printbuf, event_time->tv_usec);
+        printf (option_json ? "{\"timestamp\":\"%s.%06li\"" : "%s.%06li ", printbuf, event_time->tv_usec);
     } else if (option_timestamp == 2) {
-        printf ("%li.%06li ", event_time->tv_sec, event_time->tv_usec);
+        printf (option_json ? "{\"timestamp\":%li.%06li" : "%li.%06li ", event_time->tv_sec, event_time->tv_usec);
     }
 
     /* print user and group */
     if (option_user && proc_fd_stat.st_uid != (uid_t)-1)
-        snprintf(printbuf, sizeof printbuf, " [%u:%u]", proc_fd_stat.st_uid, proc_fd_stat.st_gid);
+        snprintf(printbuf, sizeof printbuf,
+                 option_json ? ",\"uid\":%u,\"gid\":%u" : " [%u:%u]",
+                 proc_fd_stat.st_uid, proc_fd_stat.st_gid);
     else
         printbuf[0] = '\0';
 
-    printf ("%s(%i)%s: %-3s %s\n", procname[0] == '\0' ? "unknown" : procname, data->pid, printbuf, mask2str (data->mask), pathname);
+    if (option_json) {
+        putchar(option_timestamp ? ',' : '{');
```

Comment:
Can you please simplify this? Always print the opening `{` unconditionally, and always append the `,` to the timestamp printing above.